### ISSUE TITLE
Azure App Service Deploy fails for .net core 2.0 with ERROR_FILE_IN_USE

### ISSUE THREAD

**Issue by IvanAlekseev** [Microsoft/azure-pipelines-tasks#5259]
I am trying to deploy an Asp.Net Core 2.0 web app to Azure App Service and receive this error 
```
2017-09-07T15:20:07.7379097Z ##[section]Starting: Deploy Azure App Service
2017-09-07T15:20:07.7569101Z ==============================================================================
2017-09-07T15:20:07.7569101Z Task         : Azure App Service Deploy
2017-09-07T15:20:07.7569101Z Description  : Update Azure Web App Services, Web App On Linux , Function Apps, Mobile Apps using Web Deploy / Kudu REST APIs
2017-09-07T15:20:07.7569101Z Version      : 3.3.14
2017-09-07T15:20:07.7569101Z Author       : Microsoft Corporation
2017-09-07T15:20:07.7569101Z Help         : [More Information](https://aka.ms/azurermwebdeployreadme)
2017-09-07T15:20:07.7569101Z ==============================================================================
2017-09-07T15:20:11.9902809Z Got connection details for Azure App Service:'██████████████'
2017-09-07T15:20:16.5780995Z [command]"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package='d:\a\r1\a\██████████████\drop\██████████████.zip' -dest:auto,ComputerName='https://██████████████.scm.azurewebsites.net:443/msdeploy.axd?site=██████████████',UserName='********',Password='********',AuthType='Basic' -setParam:name='IIS Web Application Name',value='██████████████' -enableRule:AppOffline -retryInterval:6000 -retryAttempts:10 -enableRule:DoNotDeleteRule -userAgent:VSTS_e29e8f6a-6b5e-463d-ba6e-72dd7866010e_release_10_50_99_1
2017-09-07T15:20:17.1910281Z Info: Using ID '████████████████████████████████████' for connections to the remote server.
2017-09-07T15:20:21.1058805Z Info: Using ID '████████████████████████████████████' for connections to the remote server.
2017-09-07T15:20:27.0283815Z Info: Updating file (██████████████\appsettings.Development.json).
2017-09-07T15:20:27.0283815Z Info: Updating file (██████████████\appsettings.json).
2017-09-07T15:20:27.0283815Z Info: Updating file (██████████████\Connected Services\Application Insights\ConnectedService.json).
2017-09-07T15:20:27.0283815Z Info: Updating file (██████████████\██████████████.Contract.dll).
2017-09-07T15:20:27.0283815Z Info: Updating file (██████████████\██████████████.Contract.pdb).
2017-09-07T15:20:27.0283815Z Info: Updating file (██████████████\██████████████.deps.json).
2017-09-07T15:20:27.0283815Z Info: Updating file (██████████████\██████████████.dll).
2017-09-07T15:21:24.7852633Z ##[error]Failed to deploy web package to App Service.
2017-09-07T15:21:24.7862628Z ##[warning]Try to deploy app service again with Rename locked files option selected.
2017-09-07T15:21:24.7862628Z ##[error]Error Code: ERROR_FILE_IN_USE
More Information: Web Deploy cannot modify the file '██████████████.dll' on the destination because it is locked by an external process.  In order to allow the publish operation to succeed, you may need to either restart your application to release the lock, or use the AppOffline rule handler for .Net applications on your next publish attempt.  Learn more at: http://go.microsoft.com/fwlink/?LinkId=221672#ERROR_FILE_IN_USE.
Error count: 1.

2017-09-07T15:21:24.7862628Z ##[error]Error: C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe failed with return code: 4294967295
2017-09-07T15:21:26.2210992Z Successfully updated deployment History at https://██████████████.scm.azurewebsites.net/deployments/501504797684782
2017-09-07T15:21:26.2321177Z ##[section]Finishing: Deploy Azure App Service
```

I've tried some workarounds from similar issue without any luck https://github.com/Microsoft/vsts-tasks/issues/1607

Any ideas on what should I do to fix this?

**Comment by vincentdass** [comment]
@IvanAlekseev , Can you please share the full logs for the release?

**Comment by IvanAlekseev** [comment]
@vincentdass added more info. Will it be enough?

**Comment by vincentdass** [comment]
@IvanAlekseev , Is there any web jobs enabled in App Service (including Application Insights)?
Did you enable `Rename locked files` option too?

**Comment by IvanAlekseev** [comment]
@vincentdass - this is a dummy service name - http://errorfileinuse-msexample.azurewebsites.net and we have a problem with recommendations api.
We have app insights enabled and no other webjobs.
I will try with `Rename locked files` option and tell you the results, for now we deploy to staging and then swap with prod

**Comment by nphmuller** [comment]
Same issue here: [dummy](dummy-fileinuse.azurewebsites.net) Happens on both the staging and production variant of the api app. I have `Rename locked files` and `Take App Offline` enabled. Still get the error.

Retrying the deployment usually works.

I have tried `-retryAttempts:10` but not `retryInterval:6000`. Will probably work as a workaround when I try the deployment with both.

**Comment by pan-wang** [comment]
could you please collect event log from AppService portal and  share it? In general, a windows event will be logged once the file change of app_offline.htm is received and the IIS component try to gracefully shutdown the backend asp.net process.

**Comment by jeffputz** [comment]
There are at least three versions of this issue floating around, all with workarounds that don't work for everyone. It became an issue again with .NET Core v2 release time... perhaps the thing that changed has something to do with it? Having to change your build process and endure a lot of pain with every release isn't great for productivity.

**Comment by rmarinho** [comment]
Ok let's just talk here :) i have the same issue, workaround is manual turn off the slot.. i don't like that kills productivity ..

**Comment by hkusulja** [comment]
It worked for me before, and it stopped working about 2 months ago, and still have issues. My app is asp.net core v1 including the asp.net 4.6.1 , I did not change my app or deployment procedure. Issue still persist. Second build always passes. I have same issue for multiple VS TS projects / and azure web sites.

**Comment by vincentdass** [comment]
@hkusulja , Is App offline flag enabled?
Can you please try adding `-retryAttempts:10  -retryInterval:6000` in Additional arguments of Web Deploy?
In this way, we increase the retry time for deployment.

**Comment by jeffputz** [comment]
I did that and it had no effect.

**Comment by TroySchmidt** [comment]
Having the same issue.  Using a continuous deployment release definition in VSTS.  It randomly fails deploying because the file is still in use.  I have tried the retryAttemps, retryInterval, put stop and start management tasks around the deploy.  This is still happening regardless of any workarounds other than manually redeploy until it works.

**Comment by tomkerkhove** [comment]
Same issue here

**Comment by ColinM9991** [comment]
Having the same issue on an Azure App Service environment with no fix on the retryAttempts and retryInterval.

It works fine with "app_offline.htm", but not "App_Offline.htm".
This is apparently fixed in a later version of the ASP.NET Core Module for IIS, but this isn't something that can exactly be installed on an Azure App Service node.

**Comment by vincentdass** [comment]
@davidebbo, @pan-wang , to comment on the issue.
The casing issue in app_offline is fixed long back.

**Comment by pan-wang** [comment]
@ColinM9991 the casing issue was fixed this Jan. When did you hit this casing issue last time?

**Comment by ColinM9991** [comment]
@PanWang, it was hit both on Sunday, and yesterday (Monday).

**Comment by fmdufour** [comment]
@pan-wang  I'm hitting this issue right now.

**Comment by pan-wang** [comment]
@fmdufour could you please collect server side  windows event log (during the deploy) through portal and share it ?

**Comment by ColinM9991** [comment]
@pan-wang 

I've retried following an environment rebuild and it seems to be working now, I will put it under a bit of load once I am at home again and attempt a redeployment.

**Comment by fmdufour** [comment]
@pan-wang  Are you refering to eventlog.xml file? I couldn't find any relevant information on it.

**Comment by TroySchmidt** [comment]
I updated to .NET Core v1.1.4 and that seems to have fixed it.  It does appear that I was having the casing issue otherwise.

**Comment by hoetz** [comment]
Hit this today as well with a aspnetcore 2.0 azure web app and I even enabled the rename locked file option. Thing is it worked at some point but now it doesn't.

**Comment by vincentdass** [comment]
@hoetz , Can you please add  `-retryAttempts:10 -retryInterval:6000`  in additional arguments of Web Deploy options and rerun the release?
Please share the debug logs for the same.

**Comment by nphmuller** [comment]
@vincentdass while this worked for me, this _is_ seen as a workaround, right? 
Are you still interested in the debug logs, even with this workaround implemented? The issue didn’t occur every deployment for me, so I’m not sure if the log is still useful with the workaround.
Got an email address I can mail confidential logs to?

**Comment by vincentdass** [comment]
@nphmuller , thanks for the response.
Wanted to confirm from the logs, if the issue is raised by any Web jobs or DLL.
Can you please share only the ASPCore version?

**Comment by hoetz** [comment]
Setting -retryAttempts:10 -retryInterval:6000 seemed to work for now

**Comment by PaitoAnderson** [comment]
Still getting this error occasionally, anything else I can try?

It really bad because it takes down our site and says "Site Under Construction" until I can click "New Build" in AppVeyor and it completes :(

I am using AppVeyor with these deploy options:
```
- provider: WebDeploy
  server: https://XXX.scm.azurewebsites.net:443/msdeploy.axd?site=XXX
  website: XXX
  username: $XXX
  password:
    secure: XXX
  aspnet_core: true
  app_offline: true
  sync_retry_attempts: 10
  sync_retry_interval: 6000
  remove_files: true
```

Error:
```Web Deploy cannot modify the file 'XXX.exe' on the destination because it is locked by an external process.  In order to allow the publish operation to succeed, you may need to either restart your application to release the lock, or use the AppOffline rule handler for .Net applications on your next publish attempt.  Learn more at: http://go.microsoft.com/fwlink/?LinkId=221672#ERROR_FILE_IN_USE.```

**Comment by pan-wang** [comment]
@PaitoAnderson  you hit "Site Under Construction" is because web deploy dropped app_offline.htm file which is to signal ANCM to shutdown the backend process so that Web Deploy can overwrite the targeted assemblies. ANCM should have already sent shutdown signal to the backend. Could you please share the windows event log from the server

**Comment by vhavman** [comment]
@vincentdass Hi Vincent, we are facing the same issue which is going on in this thread, I am getting FILE_IN_USE error even after trying the following approach.
1. 
![image](https://user-images.githubusercontent.com/32943935/31798074-9ed2cb0c-b4e6-11e7-832a-d60926b3a012.png)

2. 
![image](https://user-images.githubusercontent.com/32943935/31798096-b94c7d52-b4e6-11e7-9401-3018abd88ce6.png)

3. 
![image](https://user-images.githubusercontent.com/32943935/31798119-d6862396-b4e6-11e7-982d-86f4dbdbdfe3.png)

Now all these are in place for my deployment process but still getting same error. Please let me know if anything is missing or am i doing anything in wrong way.

Thanks in advance.

**Comment by jeffputz** [comment]
This sure seems like a common problem for there not to be any solution, or feedback from the VSTS team to at least say, "We're working on it."

**Comment by mmulhearn** [comment]
Also seeing this on multiple deployments.

"Web Deploy cannot modify the file 'VBCSCompiler.exe' on the destination because it is locked by an external process"

**Comment by pan-wang** [comment]
Is .net core application? Was app-offline option enabled from the deployment portal?

**Comment by mmulhearn** [comment]
No, it's a framework web app.  

No, it was not.

**Comment by PaitoAnderson** [comment]
@pan-wang It's an App Service, is Event Log an option in Kudu or something?

Also, for the last 11 days I have been using this option in AppVeyor:

`aspnet_core_force_restart: true`

and I haven't hit a file lock since, could be a coincidence I am not sure.

**Comment by dylan-smith** [comment]
FYI, I'm getting this on a pretty regular basis (every other deploy or so), and my app is NOT core.

ASP.Net WebApi 2 (not aspnet core), deploying to azure app service.  Usually the file in use is VBCSCompiler.exe

Happy to let you poke around in my account (dylanfromwinnipeg "at" gmail.com)

**Comment by vincentdass** [comment]
@dylan-smith , Is the issue happening with App Offline enabled?
Can u add `-retryAttempts:10 -retryInterval:6000` in additional arguments of web deploy option?

**Comment by jeffputz** [comment]
The extra arguments make no difference for me.

**Comment by dylan-smith** [comment]
@vincentdass I had app offline, and the rename file thing enabled.  I added the retry args and it didn't fix anything.

**Comment by davidebbo** [comment]
@dylan-smith please see https://github.com/projectkudu/kudu/wiki/Dealing-with-locked-files-during-deployment#for-msdeploy-try-enabling-app-offline, and you the steps to determine if you're in case 1 (app_offline.htm is ignored) or in case 2 (it's taking a while).

**Comment by pan-wang** [comment]
I think the problem might be that Roslyn compiler (VBCSCompiler.exe) was running as out-of-process (This is a new compiler feature compared with the old in-process csc compiler). 
App_offline won't help for this scenario. App_offline only unloads app domain for asp.net application but cannot kill the child Roslyn process. This is why user still hit the file still in use issue even after enabling app_offline.  
User has to fully stop the the application (IIS application pool) before deploying the new VBCSCompiler.exe.

**Comment by davidebbo** [comment]
I would still expect VBCSCompiler.exe to shut down eventually. @dylan-smith please do go through the steps I highlighted above which will give us useful data.

**Comment by IvanAlekseev** [comment]
Having this configuration in asp.net core 2.0 project works fine for us now

![image](https://user-images.githubusercontent.com/23052247/32619980-43e6b610-c574-11e7-95a8-5574a92be0d1.png)

**Comment by hkusulja** [comment]
We did not change anything, and problem started about 1-2 months ago, but it recently stopped occurring, so from my side it works. (same asp.net core 1 and 2 projects)...,

**Comment by frankfuu** [comment]
Having the same problem here

```
2017-11-19T16:20:25.5530009Z ==============================================================================
2017-11-19T16:20:25.5530009Z Task         : Azure App Service Deploy
2017-11-19T16:20:25.5530009Z Description  : Update Azure WebApp Services On Windows, Web App On Linux with built-in images or docker containers, ASP.NET, .NET Core, PHP, Python or Node based Web applications, Function Apps, Mobile Apps, Api applications, Web Jobs using Web Deploy / Kudu REST APIs
2017-11-19T16:20:25.5530009Z Version      : 3.3.25
2017-11-19T16:20:25.5530009Z Author       : Microsoft Corporation
2017-11-19T16:20:25.5530009Z Help         : [More Information](https://aka.ms/azurermwebdeployreadme)
2017-11-19T16:20:25.5530009Z ==============================================================================
```

```
2017-11-19T16:21:35.9113361Z Info: Updating file (myapp123\bin\roslyn\csc.exe).
2017-11-19T16:21:36.1193239Z ##[error]Failed to deploy web package to App Service.
2017-11-19T16:21:36.1213241Z ##[warning]Try to deploy app service again with Rename locked files option selected.
2017-11-19T16:21:36.1213241Z ##[error]Error Code: ERROR_FILE_IN_USE
More Information: Web Deploy cannot modify the file 'csc.exe' on the destination because it is locked by an external process.  In order to allow the publish operation to succeed, you may need to either restart your application to release the lock, or use the AppOffline rule handler for .Net applications on your next publish attempt.  Learn more at: http://go.microsoft.com/fwlink/?LinkId=221672#ERROR_FILE_IN_USE.
Error count: 1.

2017-11-19T16:21:36.1233254Z ##[error]Error: C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe failed with return code: 4294967295
2017-11-19T16:21:37.1842618Z Successfully updated deployment History at https://*****.scm.azurewebsites.net/api/deployments/2451511108496095
```

**Comment by deanilvincent** [comment]
I have the same issue. I'm using .net core 2 and in the first cd, it works fine but when I've tried to update a simple file, then all the following releases are showing the same error
`
...
ERROR_FILE_IN_USE
...
`

**Comment by vincentdass** [comment]
@frankfuu , Can you please make below changes to your task?
![image](https://user-images.githubusercontent.com/20121508/33010956-5f8b7ce8-ce02-11e7-92b1-f6cf81f7a684.png)

If you are still facing the issue, we suggest you to stop the app service, deploy and start the app service.
For zero down-time deployment, you can deploy to your staging slot and swap it with production.
`Azure App service Manage` task can be handy here.

**Comment by frankfuu** [comment]
Hey @vincentdass, I'll give it ago, but I can't consistently reproduce the issue. We already have slot swapping for our production environments so it's ok. I just want this to be implemented for our UAT/CI environment without a swapping mechanism. I'm just hoping the `-retryAttempts:10 -retryInterval:6000` will do the trick

**Comment by niadak** [comment]
@frankfuu , Did you get a chance to try the above suggestion? Are you still facing the issue?

**Comment by frankfuu** [comment]
hi @niadak Yep, it seems to be working so far! I'm hoping to kind of go back and see how many times it actually did retry.

**Comment by DaleMckeown** [comment]
It should be noted that this issue is not limited to Azure - publishing to IIS via Web Deploy causes the same issues.

**Comment by buzallen** [comment]
This is still an issue for me, but only on certain projects and I can't find a common thread as of this point. This is a .net core 2.0 MVC application being published to an azure web app from visual studio using the publishing profile. Only way around it is to stop the service and then restart it. Note this is not a slot in this case, it's the regular production app.

Over the summer this was an issue on some other projects but that went away and I assumed the problem was fixed. In two new projects this issue is back. I have the retry attempts set at 20 but per another issue thread that is not needed anymore.  

I'd guess about 50% of publishes fail with this error.  I know I can use a staging slot and swap to make it work and on our primary app I do that but there are some smaller projects I'd really like to just be able to publish simply.

**Comment by vincentdass** [comment]
@buzallen , Can you please provide the [details](https://github.com/Microsoft/vsts-tasks/issues/5259#issuecomment-329572378) asked by @pan-wang , so we can closely debug this issue?

**Comment by buzallen** [comment]
@vincentdass I read the @pan-wang  post about event logs, is it the eventlog.xml file in the 'Logs' directory?  I've enabled web server logging for the app but the eventlog.xml file doesn't appear to have any errors in it even though a publish failed after I enabled logging.

If there are any details on how to capture the specific logs, or if I am not getting the correct data please let me know.  FWIW, I was able to publish without error twice before it failed the third time.

**Comment by pan-wang** [comment]
@buzallen ANCM logs failure to Windows Event log. You can try the method posted at http://www.c-sharpcorner.com/article/azure-app-service-troubleshoot-using-event-log/ to get the windows event log.

**Comment by pgmolloy** [comment]
Getting this issue today, I had it a while back. Have to do the same thing stop the service, deploy and restart. Here's the info.

2017-12-31 4:44:09 PM
System.AggregateException: One or more errors occurred. ---> System.Exception: Publish failed due to build errors. Check the error list for more details.
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Web.Publish.PublishService.VsWebProjectPublish.<>c__DisplayClass40_0.<PublishAsync>b__2()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.ApplicationCapabilities.Publish.ViewModel.ProfileSelectorViewModel.<RunPublishTaskAsync>d__108.MoveNext()
---> (Inner Exception #0) System.Exception: Publish failed due to build errors. Check the error list for more details.<---

===================



Severity	Code	Description	Project	File	Line	Suppression State
Error		Web deployment task failed. (Web Deploy cannot modify the file 'rsocapiservice.dll' on the destination because it is locked by an external process.  In order to allow the publish operation to succeed, you may need to either restart your application to release the lock, or use the AppOffline rule handler for .Net applications on your next publish attempt.  Learn more at: http://go.microsoft.com/fwlink/?LinkId=221672#ERROR_FILE_IN_USE.)	0

**Comment by vincentdass** [comment]
@pgmolloy , the recommendation from App Service team is to stop, deploy and start the app (as you stated), when the app_offline doesnt have any effect (even with retry-attempts and retry-interval).

**Comment by nordvall** [comment]
I've been doing almost a hundred deployments over the last months and never hit this issue. I am deploying Azure Functions and ASP.NET Core 2.0 apps with VSTS Release Management.

However today this issue appeared on one of my apps (ERROR_FILE_IN_USE for my application dll). One special thing about this app is that it was manually stopped from the Azure Portal during the Christmas holidays. Yesterday I did a new deployment, and then manually started the app again from the Azure Portal (when I realized it was stopped). Everything was then working again, but today the deployments starts failing. I have no special parameters configured in the deployment task (since I have not needed any tweaks). 

I have no idea about the inner workings of this, but I am thinking that maybe manually starting and stopping apps in some way interfere with the web deployment's ability to release file locks?

I've now surrounded the deployment task with "Azure App Service Manage" start/stop tasks in the process and the deployments now works again.

**Comment by vincentdass** [comment]
For FILE_IN_USE issue, There are few recommendation that resolves the issue.

- Enable `Take App Offline` option. With `-retryAttempts:10 -retryInterval:6000` additional argument provides a reliability with retry.
- Enable `Rename locked files` option to allow MSDeploy to attempt to rename DLLs if they can't be copied during deployment
The above two options works for most of the cases.

![image](https://user-images.githubusercontent.com/20121508/34953009-a0095b30-fa41-11e7-8216-a680b4d1aa30.png)

If the issue is still not resolved, try stopping the app service, deploy and then start the app service.
Swap slot helps in Zero-down time deployment.

Refer here for more [details](https://github.com/projectkudu/kudu/wiki/Dealing-with-locked-files-during-deployment).

Rever back if the recommendation is not working.

**Comment by vincentdass** [comment]
Closing the thread as the recommendation provided above should have resolved the issue.
Feel free to re-open the thread or new one if the issue is still persisting.

**Comment by onionhammer** [comment]
This is definitely a hack/workaround, not a very satisfying solution.

**Comment by DaleMckeown** [comment]
This also doesn't cater for people deploying to a regular IIS site, as
reported above.

Stopping an app pool, publishing a service, then restarting an app pool is
not a good process for people who have to publish an app to a single server.

**Comment by rmarinho** [comment]
Please reopen, stopping the website is not what one should be recommend to do.. We have production apps depending on 100% uptime.

**Comment by vincentdass** [comment]
Reopening the thread.
@onionhammer , @rmarinho , Is App_offline and Rename locked files option not resolving your issue?
@DaleMckeown , Are you referring to IIS App Deploy task?

**Comment by onionhammer** [comment]
@vincentdass It works around the core issue. A workaround is a temporary solution to a more fundamental problem. The fundamental problem is by default this task fails without a lot of monkeying around with the settings. Do you expect users in the future to simply search for a solution, find this post, and apply your workaround whenever people come across this in the future?

**Comment by vincentdass** [comment]
@onionhammer , I agree with your concern. We are currently working on an experience to enable the app offline feature and rename locked files options by default with retry attempts for MSDeploy in the release templates. Need to think through on enabling these options by default in the task, as App offline flag brings the app down for a time-interval and not accepted by others.

Now on the core problem, where the task failing without retrying with the above two options. We explicitly want to point out what went wrong and the recommendation is provided in the error logs as warning message.

`2017-09-07T15:21:24.7862628Z ##[warning]Try to deploy app service again with Rename locked files option selected.`

`More Information: Web Deploy cannot modify the file '██████████████.dll' on the destination because it is locked by an external process.  In order to allow the publish operation to succeed, you may need to either restart your application to release the lock, or use the AppOffline rule handler for .Net applications on your next publish attempt. `

**Comment by DaleMckeown** [comment]
@vincentdass Yep.

**Comment by hakankaraduman** [comment]
I'm having the same problem on my .net core 2 application. It was okay yesterday but today I have to first stop the app service and then I can upload and restart the app service.

What's the workaround for this? I publish from the visual studio 2017 community edition.

**Comment by jonathaneckman** [comment]
I got stuck behind the same issue deploying my .NET Core 2 app to an app service using every variety of work around in this thread. 

Make sure you queue a new release if you change your release task. Using redeploy will not pick up the changes suggested above.

**Comment by PaitoAnderson** [comment]
I wonder if we could use the new Run-From-Zip deployment option to solve these problems?

Azure/app-service-announcements#84

**Comment by davidebbo** [comment]
@PaitoAnderson it's not quite ready for prime time, but I do think it will provide a solution to many of these issues.

**Comment by Korporal** [comment]
@davidebbo @IvanAlekseev 

I've just encountered this problem after my first experiments in publishing .Net Core Web Apps to Azure. The problem is intermittent for me. I published a simple app this morning and (other than unrelated problems https://github.com/aspnet/JavaScriptServices/issues/1330 and https://github.com/aspnet/JavaScriptServices/issues/1282) had no problem. I published the app several times over the course of an hour as I made tiny tweaks to it (cosmetic).

I then went out for an hour, came back made another small tweak and tried to publish whereupon I hit this locked assembly issue.

I stopped the app in azure and it then published fine, I made a small change and published without stopping it and that too went fine.

I have to say, the presence of this problem strikes me as rather incredible given the fundamental nature of what we're discussing here. If publishing to Azure using the latest MS tools fails like this then what is going on? This issue seems to have been around for quite a while and it's not being treated as a priority 1 bug?

I have no idea what goes on under the hood with a publish, but surely part of the protocol must involve a request to stop the app as one of the first steps?

I'm interested to know too if the problem is in VS 2017 or Azure, does anyone at least know this much?

Thanks.

**Comment by pan-wang** [comment]
I recently got a local repro with VS publishing(not on azure). What I noticed is that the published was trying to delete the whole app folder which is under file watching by ANCM for app_offline change and IIS for configuration change.  All files in the target folder is marked as deleting and publishing failed.  Interesting thing is that process monitor did not report any other process tried to access files except the publisher. I coulf manually copy and delete files/folders for the same scenario. Will work with VS publisher team to see how we can fix it.

**Comment by vincentdass** [comment]
@pan-wang , any updates on the issue?

**Comment by pan-wang** [comment]
@vincentdass  Webdeploy made some change to avoid race condition due to file watcher. I have not try the new bits to see whether it also solve our problem. Will do it.

**Comment by mdsharpe** [comment]
Frequently getting this problem deploying an asp.net core 2.0 web app to Azure using Octopus Deploy. Often have to manually retry many times. Seems intermittent.

**Comment by BowserKingKoopa** [comment]
I'm having the same problem. I'm using the Publish option in Visual Studio 2017 for an ASP.NET Core on Core app. This has never happened to me before during my years of clicking Publish for classic ASP.NET apps. For my ASP.NET Core app this happens all the time. I can get around it by shutting off the app from Azure Portal. This needs a real fix, Publish is broken.

**Comment by pan-wang** [comment]
@BowserKingKoopa did you enable app-offline for deployment? as there is no shallow copy for core app, appoffline must be used to shutdown the kestrel application. if possible, could you please collect and share the windows event log during deployment? this is log will tell whether the file change notification was received and what the response received from backend for shutting down.

**Comment by Korporal** [comment]
This is hitting people pretty frequently so how about Microsoft making an effort to reproduce it and doing the right thing with this? its close to a "show stopper" for some of us.

**Comment by DaleMckeown** [comment]
@Korporal Agreed.

It's getting to the point where I'm actively putting off publishing updates to sites for my organisation now.

Overall, I love AspNetCore, but publishing sites has been a significant step backwards in my opinion.

**Comment by PaitoAnderson** [comment]
I've stopped using MSDeploy with AspNetCore because of this issue.

I tried Run-From-Zip (Preview), but there are AspNetCore issues with it as well currently (Documented here: Azure/app-service-announcements-discussions#32)

I ended up using ZipDeploy, but I have to manually click "Restart" on the App Service after every deployment to get it to pick up the changes.

**Comment by davidebbo** [comment]
@PaitoAnderson (and others) I have some potentially good news for things to come in the next week or so:

- The Run-From-Zip issue is being addressed. The fix is already in some places but not yet everywhere,
- We're making changes to ZipDeploy such that if the app is set up for Run-From-Zip (using `WEBSITE_USE_ZIP=1`, not the external URL), ZipDeploy changes its behavior to use the passed in zip file for Run-From-Zip (instead of just expanding it to `wwwroot` is it normally would - `wwwroot` is read-only in this case).

I know this sounds a little complicated, but bottom line is that in theory, to get things working with .NET Core and VSTS, all you have to do is:

- `WEBSITE_USE_ZIP=1`
- Tell vsts to use zipdeploy (and not use app_offline)

Does that make sense?

**Comment by PaitoAnderson** [comment]
I tried added WEBSITE_USE_ZIP=1 on two different App Services today both in East US, (1) Is targeting `netcoreapp2.0` using AspNetCore, no problems, zip deployment worked fine too! (2) Is targeting `net471` using AspNetCore, instant 500 errors when applying the application setting, any ideas?

**Comment by davidebbo** [comment]
@PaitoAnderson unfortunately, there were some delays, and the deployment in East US is only partial. To confirm that it is the reason, please try the following:

- For each of the two apps, go to Kudu (AppName.scm.azurewebsites.net)
- Look at the Azure App Service field (not the 'Build'). 71.0.x is the old build, and 72.0.x is the new one

Does that match your working / non-working app?

**Comment by PaitoAnderson** [comment]
Yes, that matches the working / non-working, I'll try again when that updates. Thank you!

**Comment by davidebbo** [comment]
@PaitoAnderson Chances are you got the update now, so please give it a try when you have a chance.

**Comment by PaitoAnderson** [comment]
@davidebbo I tried my best, but I hit different issues, maybe it's me this time.

Zip Deployment: I set WEBSITE_USE_ZIP=1 and I didn't get 500 errors, cool! Next, I did a zip deployment but it didn't pickup my updated dlls (C# changes), html/javascript were new however.

Run-From-Zip: I deployed to a Azure Blob, got the SAS URL (I tested downloading the zip using a browser), I set WEBSITE_USE_ZIP to the SAS URL but it just started returning 403 "You do not have permission to view this directory" errors.

**Comment by davidebbo** [comment]
@PaitoAnderson can you email me details (david.ebbo (at) microsoft.com), including the test site name and the zip url? We'll figure out what's going on.

**Comment by andreujuanc** [comment]
I have an Azure function deployed from VS for testing purposes and I'm having the same issue.
And I cant use the renaming flag because i need to update the locked file to a new version.

It's so meh to go to the portal and stop > publish > start every time i publish. 

I'm just testing a solution via VS publish profile, havent configure this solution to our CD pipeline.

Any tips will be appreciate it.
PS: already have  `<EnableMsDeployAppOffline>true</EnableMsDeployAppOffline>`
`
C:\Program Files\dotnet\sdk\2.1.104\Sdks\Microsoft.NET.Sdk.Publish\build\netstandard1.0\PublishTargets\Microsoft.NET.Sdk.Publish.MSDeploy.targets(139,5): error : Web deployment task failed. (Web Deploy cannot modify the file 'DurableTask.AzureStorage.dll' on the destination because it is locked by an external process.  In order to allow the publish operation to succeed, you may need to either restart your application to release the lock, or use the AppOffline rule handler for .Net applications on your next publish attempt.  Learn more at: http://go.microsoft.com/fwlink/?LinkId=221672#ERROR_FILE_IN_USE.) [C:\Users\andre\Source\Repos\WgmCodeJobs\DurableFunctions\WgmCodeJobs\WgmCodeJobs.csproj]
`

**Comment by jsheetzati** [comment]
@davidebbo What is the recommended way to "Tell vsts to use zipdeploy (and not use app_offline)"? Is there a way to do this with the Azure App Service Deploy task?

**Comment by SumiranAgg** [comment]
@jsheetzati VSTS consuming zipdeploy is work in progress and we will let you know once it is ready.

**Comment by thomasrousset** [comment]
Hi,

so to use the fix we have to add to the appsettings' webapps this key WEBSITE_USE_ZIP=1 ; that's it ?

**Comment by davidebbo** [comment]
@thomasrousset You also need to you zipdeploy for deployment. See my previous comment. Also, note that `WEBSITE_USE_ZIP` was renamed to `WEBSITE_RUN_FROM_ZIP`.

**Comment by PrimeHydra** [comment]
Was hitting this issue trying to deploy a .NET Core 2.0 Azure Function. Tried WEBSITE_USE_ZIP=1, didn't help. Tried checking Take App Offline, no joy.

What worked for me: Checking "Publish Using Web Deploy" and "Rename Locked Files", and creating a new deployment. Like jonathaneckman said, Redeploy won't work because it won't pick up your fixes to the release definition!

**Comment by PrimeHydra** [comment]
@davedboo no, I removed the WEBSITE_USE_ZIP=1 and did not replace it. Simply checking Use Web Deploy and Rename Locked Files did the trick for me. I'm about to try this setting on the rest of my deploying services and see if they all deploy cleanly.

PS - I agree with the sentiment expressed by others, that on shouldn't have to muck with default settings just to work around issues like locked files. I never had to change anything when deploying from Visual Studio, for it to work perfectly.

**Comment by davidebbo** [comment]
What I meant was in response to you saying

> Tried WEBSITE_USE_ZIP=1, didn't help.

If at the time you tried it you did not also use zipdeploy, that attempt could not have possibly worked. The fact that you removed it after the fact to try something else is unrelated to my question.

Run-From-Zip is still a new thing, which is why tooling support is not yet there.

**Comment by PrimeHydra** [comment]
@davidebbo I only joined this thread to post a solution in hopes it would help someone else. If by "it has no chance of working" you meant "it has no chance of using zip deploy", then you may be right. But the only definition of "working" that I care about is that it deployed without errors.

**Comment by davidebbo** [comment]
@PrimeHydra Understood. Thanks for working what worked for you.

I just wanted to make sure that your comment about `WEBSITE_RUN_FROM_ZIP` would not be interpreted by others as an issue with that feature (which it isn't).

**Comment by thomasrousset** [comment]
@davidebbo hi and thanks for your answer.

What I did is to check 'publish usign webdeploy', 'Remove additional files at destination' and 'Rename locked files'. On the webapp's side I add the appsettings  with key : WEBSITE_RUN_FROM_ZIP and value : 1.

> However I still have the file in use error : More Information: Web Deploy cannot modify the file 'AF'xxx.exe' on the destination because it is locked by an external process.  In order to allow the publish operation to succeed, you may need to either restart your application to release the lock, or use the AppOffline rule handler for .Net applications on your next publish attempt.  Learn more at: http://go.microsoft.com/fwlink/?LinkId=221672#ERROR_FILE_IN_USE.
> Error count: 1.

**Comment by davidebbo** [comment]
@thomasrousset `WEBSITE_RUN_FROM_ZIP` has no chance of working unless you deploy using zipdeploy. Please see my previous responses.

**Comment by PrimeHydra** [comment]
@thomasrousset you'll also need to create a new release from your saved definition. (If you re-deploy an existing release it will use the older release definition.) That tripped me up once, just though I'd mention it.

I don't know if "Remove additional files at destination" might also be trying to delete locked files, maybe uncheck that as well.

Beyond that, I'm out of ideas--good luck! :)

**Comment by BowserKingKoopa** [comment]
Zip deploy sounds like the solution to this, as well as just a simpler cleaner way to deploy apps in general. I'll be sitting tight until Visual Studio supports it.  Did we ever get an answer to why this is broken in Core but always worked before in previous version of asp.net and publish?

**Comment by davidebbo** [comment]
@BowserKingKoopa it's because Core loads assemblies in-place, while regular ASP.NET shadow copies everything (so they're not locked in-place). So in Core, it relies on App_Offline.htm to unload things, and evidently that does not work reliably enough.

**Comment by DaleMckeown** [comment]
@BowserKingKoopa Not that I'm aware of. I'm waiting for @pan-wang to report his ANCM findings based on his repro.

**Comment by thomasrousset** [comment]
@davidebbo may you show me where do you set the 'using zipdeploy' option please ?

**Comment by davidebbo** [comment]
@thomasrousset: @vincentdass or others working on VSTS can better answer this (I'm on the App Service side).

**Comment by vincentdass** [comment]
@thomasrousset , we have added this feature in current sprint and will be available in next deployment of 4.* task version.
As a workaround, you can use the curl command provided [here ](https://github.com/projectkudu/kudu/wiki/Deploying-from-a-zip-file)to use the Zip Deploy command.
You can also pull from the repo and upload the task to your account.

**Comment by thomasrousset** [comment]
@vincentdass ok thank you for your answer. When do you plan to publish it ? Is it possible to share it whith us in private mode ?

**Comment by frankfuu** [comment]
@thomasrousset, I'm also waiting for this feature.

I think VSTS tries to release every 3 weeks and their last release was on [19th June
](https://docs.microsoft.com/en-us/vsts/release-notes/#current-features) so i'm hoping it would be around [10th July](https://www.timeanddate.com/date/dateadded.html?d1=19&m1=06&y1=2018&type=add&ay=&am=&aw=3&ad=&rec=)?

**Comment by vincentdass** [comment]
@frankfuu , @thomasrousset , Zip Deploy & RunFromZip feature is available in 4.* task version.
Please try out the same and share your feedback.

**Comment by frankfuu** [comment]
hi @vincentdass, just tried using Zip Deploy but it came back with an error 

> Publish using zip deploy option is not supported for msBuild package type.

Here is my msbuild command

`C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin\msbuild.exe" "C:\vsts\MyTestPool\my-test-bd-agent-2\_work\1\s\MyCompany\src\UI\My.Project\MyCompany.My.Project.csproj" /nologo /nr:false /dl:CentralLogger,"C:\vsts\MyTestPool\my-test-bd-agent-2\_work\_tasks\VSBuild_71a9a2d3-a98a-4caa-96ab-affca411ecda\1.126.0\ps_modules\MSBuildHelpers\Microsoft.TeamFoundation.DistributedTask.MSBuild.Logger.dll";"RootDetailId=313d5669-d4b5-413c-b932-1c07e05e554e|SolutionDir=C:\vsts\MyTestPool\my-test-bd-agent-2\_work\1\s\MyCompany\src\UI\My.Project\MyCompany.My.Project"*ForwardingLogger,"C:\vsts\MyTestPool\my-test-bd-agent-2\_work\_tasks\VSBuild_71a9a2d3-a98a-4caa-96ab-affca411ecda\1.126.0\ps_modules\MSBuildHelpers\Microsoft.TeamFoundation.DistributedTask.MSBuild.Logger.dll" /p:DeployOnBuild=true /p:WebPublishMethod=Package /p:PackageAsSingleFile=true /p:SkipInvalidConfigurations=true /p:PackageLocation="C:\vsts\MyTestPool\my-test-bd-agent-2\_work\1\a\\ISWebApps\\MyCompany.My.Project\\" /p:PrecompileBeforePublish=true /p:UseMerge=true /p:SingleAssemblyName=AppCode /p:WarningLevel=1 /p:platform="AnyCPU" /p:configuration="UAT" /p:VisualStudioVersion="15.0" /p:_MSDeployUserAgent="VSTS_15d68490-21eb-479b-8bf3-95d04d04d581_build_15_0"`


Does it have to be in any special format? Any documentation?

**Comment by vincentdass** [comment]
@frankfuu , Can you please edit the build as follows:

- In **Build Solution** task, change the **MSBuild Arguments** to `/p:DeployOnBuild=true /p:DeployDefaultTarget=WebPublish   /p:WebPublishMethod=FileSystem  /p:DeleteExistingFiles=True /p:publishUrl="$(System.DefaultWorkingDirectory)\\WebAppContent"`
![image](https://user-images.githubusercontent.com/20121508/43206337-1b405636-9043-11e8-98b2-b86e7d26e9d5.png)

- Add **Archive Task** and change the inputs as follows:
    - Change `Root folder or file to archive` to `$(System.DefaultWorkingDirectory)\\WebAppContent`
    - Disable `Prepend root folder name to archive paths` option
    
![image](https://user-images.githubusercontent.com/20121508/43207057-c564c380-9044-11e8-900c-dc06086fbd09.png)

**Comment by thomasrousset** [comment]
Hi,
@vincentdass I just tried the v4 with zipDeploy deployment and it works well on my side.
We'll keep on test it. Thank's for your work and the message to warn us about the v4 avaibility.

**Comment by frankfuu** [comment]
Hi @vincentdass ,

You're a champ, thanks for the detailed instructions. Got it working ;)

**Comment by vincentdass** [comment]
@pan-wang , Can you please share your findings on why App offline flag is not honored by IIS?

**Comment by pan-wang** [comment]
@vincentdass ANCM does honor appoffline. However, it depends on file change notification from OS. If FCN was not emitted or missed, nothing can be done by ANCM. This is the scenario for AppService.
In the past, WebDeploy only creates appoffline.htm file once instead of every deployment. This is why the retry logic does not help. We made change to webdeploy. That is webdeploy touches the appoffline file after deploying the appoffline and force more FCN.  The new version of Webdeploy has already been shipped and we hope it will mitigate this deployment issue.

**Comment by ssdevref** [comment]
Hi I have the same issue deploying, I have tried:
`<AppOffline>true</AppOffline>`
`<RetryAttemptsForDeployment>100</RetryAttemptsForDeployment>`
`<EnableMSDeployAppOffline>True</EnableMSDeployAppOffline>`
I have two projects: a library and the site. The library is locked when I come to do a web deploy.

**Comment by vincentdass** [comment]
PR #8457 for consuming new version of MSDeploy as per Web Deploy team recommendation.

**Comment by vincentdass** [comment]
`Run From Package` support is added in `4.*` task version of Azure App Service Deploy task.
If you still face `FILE_IN_USE` error, we recommend you to move from `Web Deploy` to `Run From Package`.

![image](https://user-images.githubusercontent.com/20121508/47496275-f51da700-d873-11e8-980d-3ef8a57b82ab.png)


[Run From Package](https://docs.microsoft.com/en-us/azure/azure-functions/run-functions-from-deployment-package#benefits-of-running-from-a-package-file) helps in avoiding the FILE_IN_USE error.
